Fix issues in the specification doc for test framework (#1291)
ballerina-platform/ballerina-spec: test/test-framework-spec.md

```
project-name/
  - Ballerina.toml
  - modules/
    -- mymodule/
      --- Module.md      
      --- mymodule.bal       
      --- resources/     
      --- tests/           <- tests for a non-default module (e.g. unit tests)
        ---- lib_test.bal  <- test file for a non-default module
        ---- resources/    <- resources for these tests
  - main.bal
  - tests/              <- tests for the default module
    -- main_test.bal    <- test file for default module
    -- resources/       <- resources for these tests

```

> In Ballerina testframework, tests are organized into per module test suites. The tests for a particular module should be added in a sub folder named `tests` within that module.

Reviewer: ```suggestion
In the Ballerina test framework, tests are organized into per-module test suites. The tests for a particular module should be added in a subdirectory named `tests` within that module.
```

Tests defined outside the test directory will not get executed when executing tests for a Ballerina package/module.

### Visibility

The symbols defined in a module are accessible from within the test files of the same module. This includes globally-defined objects and variables.
 Hence, redefining a symbol in a test file is not allowed if it is already declared in the module. Instead, they can be reassigned in the test files.

>  It must be noted that symbols defined in the test files will not be visible inside the module source files. 

Reviewer 2: ```suggestion
 Symbols defined in the test files will not be visible inside the module source files. 
```

   <td>Function specified will be run once before any of the tests in the test suite is run.
   </td>
  </tr>
  <tr>
   <td>@test:BeforeGroups {}
   </td>
    <td>Function specified will be run before any of the tests belonging to the specified groups are executed.
<p>
Annotation value fields :
<ul>

<li>groups : [“&lt;test group name&gt;”, ...] 
<ul>

<li>List of group names that this function should run before
</li> 
</ul>
</li> 
</ul>
   </td>

>   </tr>

Reviewer: We should have `serialExecution` flag for the @test:Config annotation now.

Author: Shall we add after the U9 release?

<table>
  <tr>
   <td>@test:assertEquals(any|error actual, anydata expected, string message)
   </td>
   <td>Checks if the specified value is equal to the expected value. This assertion relies on `==` based deep equality check. 
For the deep equality check to be applicable, it is required that at least one of the participating operands is of a static type that is a subtype of `anydata`. 
Consequently, the permissible scope of the expected value is confined to `anydata` type.
   </td>
  </tr>
  <tr>
   <td>@test:assertNotEquals(any actual, anydata expected, string message)
   </td>
   <td>Checks if the specified value is not equal to the expected value. This assertion utilizes the != operator for a deep equality check, thus limiting the expected value to the `anydata` type.
The actual value does not include `error` type to avoid automatically passing the tests when there is an error.

> This ensures that tests accurately reflect intended outcomes without misinterpreting errors as correct behavior.

Reviewer 2: Do we need this additional sentence?

Author: Can remove that.

   </td>
  </tr>
  <tr>
     <td>@test:assertExactEquals(any|error actual, any|error expected, string message)
     </td>

>      <td>Checks if the specified value is exactly equal to the expected value i.e. both refer to the same entity. This assertion utilizes the === operator for an exact equality check.

Reviewer 2: ```suggestion
     <td>Checks if the specified value is exactly equal to the expected value i.e. both refer to the same entity. This assertion utilizes the `===` operator for an exact equality check.
```

     </td>
    </tr>
    <tr>
     <td>@test:assertNotExactEquals
     </td>
     <td>Checks if the specified value is not exactly equal to the expected value i.e. both do not refer to the same entity.

> This assertion utilizes the === operator for an exact equality check.

Reviewer 2: ```suggestion
This assertion utilizes the `===` operator for an exact equality check.
```

     </td>
    </tr>
  <tr>
   <td>@test:assertTrue(boolean expression, string message)	
   </td>
   <td>Checks if the specified boolean expression is true.
   </td>
  </tr>
  <tr>
   <td>@test:assertFalse
   </td>
   <td>Checks if the specified boolean expression is false.
   </td>
  </tr>
  <tr>
   <td>@test:assertFail
   </td>

>    <td>Forces a test case to fail. This assertion facilitate the failing of a test during its execution when a specific condition is not met.

Reviewer 2: "This assertion facilitate the failing of a test during its execution when a specific condition is not met." - is this correct? Other assert functions also will fail the test if a specific condition is not met.

Author: Agreed that it's confusing. I'll fix it.

  Ballerina code independently from other modules and external endpoints.

### Annotation

Initializing a function mock needs a preceding annotation in order to identify and replace the occurrence of the original function during compilation.

<table>
  <tr>
   <td>@test:Mock {}
   </td>

>    <td>This annotation can be applied to either a function or a test:MockFunction intialization. When placed before a function, it identifies that function as a mock, which will be invoked whenever the original function is called.

Reviewer 2: ```suggestion
   <td>This annotation can be applied to either a function or a `test:MockFunction` intialization. When placed before a function, it identifies that function as a mock, which will be invoked whenever the original function is called.
```

Reviewer 2: Shall we mention about the mandatory functionName field also?

>  If it is a test:MockFunction initialization, the behaviour of the function needs to be stubbed in the test cases.

Reviewer 3: ```suggestion
 If it is a `test:MockFunction` initialization, the behaviour of the function needs to be stubbed in the test cases.
```

*   Mock object 

    ```ballerina
    http:Client mockClient = test:mock(http:Client, new MockHttpClient());
    ```

*   Mock function 
    

>     Compile-time mocking

Reviewer 2: We call this legacy mocking right?

Author: But is it correct to call it legacy mocking?
This is the most widely used mocking when testing services and clients right?

Author: Used the terms static and dynamic mocking.

    test:prepare(mockClient).when("functionName").doNothing();
    ```

> 3. Provide a return value. The function call will return the specified value, regardless of the arguments' values.

Reviewer 2: ```suggestion
3. Provide a return value. The function call will return the specified value, regardless of the values of the arguments.
```

    ```ballerina
    test:prepare(mockClient).when("functionName").thenReturn(5);
    ```

4. Provide a return value based on the arguments.  

    ```ballerina
    test:prepare(mockClient).when("functionName").withArguments(anydata...).thenReturn(5);
    ```

5. Mock the member variables of an object.

    ```ballerina
    test:prepare(mockClient).getMember("member").thenReturn(5);
    ```

6. Provide generalized inputs to accept any value for certain arguments.

>    This feature can be used when there is a need to selectively differentiate the mock behavior for a set of arguments.

Reviewer 2: Do we need this explanation? The spec doesn't need to explain the use case isn't it? It can be added to the learn docs if it is missing.

   ```ballerina
   test:prepare(mockClient).when("functionName").withArguments("/pets", test:ANY, ...).thenReturn(5);
   ```

7. Provide multiple return values to be returned sequentially for each function call

   This feature can be used when the arguments cannot be identified to differentiate the mock behavior based on

>    the arguments.

Reviewer 3: Can we emphasise  `withArguments` does not support with `thenReturnSequence` ?
```suggestion
   the arguments. So, `withArguments` function cannot be used with `thenReturnSequence`. It will give an error
```

   ```ballerina
   test:prepare(mockClient).when("functionName").thenReturnSequence(5, 10, 15);
   ```

> 

Reviewer 3: IMO we can introduce resource mocking here

4. Provide a return value based on the input.

    ```ballerina

>     test:when(mockFunc1).withArguments(anydata...).thenReturn(5);

Reviewer 3: ```suggestion
    test:when(mockFunc1).withArguments(anydata|error... args).thenReturn(5);
```
error also possible.. isn't it?

        test:when(intAddMockFn).withArguments(0, 0).thenReturn(-1);
        test:when(intAddMockFn).withArguments(10, 6).thenReturn(20);

        test:assertEquals(addValues(10, 6), 20, msg = "function mocking failed");
        test:assertEquals(addValues(0, 0), -1, msg = "function mocking with arguments failed");
    }

```

4. If mocking should not take place, then call the real function

```ballerina
    // main_test.bal
    import ballerina/test;
    
    @test:Mock {functionName: "intAdd"}
    test:MockFunction intAddMockFn = new ();
    
    @test:Config {}
    function testCallOriginal() {
        // Stub to call another function when `intAdd` is called.
        test:when(intAddMockFn).call("mockIntAdd");
    
        test:assertEquals(addValues(11, 6), 5, msg = "function mocking failed");
    
        // Stub to call the original `intAdd` function.
        test:when(intAddMockFn).callOriginal();
        test:assertEquals(addValues(11, 6), 17, msg = "function mocking failed");
    }
    
    // The mock function to be used in place of the `intAdd` function
    public function mockIntAdd(int a, int b) returns int {
        return (a - b);
    }

```
   

> 4. If the function does not have a return type do nothing

Reviewer 2: ```suggestion
4. If the function does not have a return type, do nothing
```

#### Test Report

In addition to the results printed in the console, a test report can be generated by passing the flag `--test-report`
 to the test execution command. The generated file is in HTML format and link to the file will be printed in
   the console at the end of test execution.

The test report contains the total passed, failed and skipped tests of the entire project.

**Example** 

```
$ ballerina test --test-report
```

### Code Coverage

The test framework provides an option to analyze the code coverage of a Ballerina project. This feature provides details about coverage of program source code by the tests executed. 

When the `--code-coverage` flag is passed to the test execution command along with the `--test-report` flag, an HTML file will be generated at the end of the test execution. The generated file is an extended version of the test report.
In addition to the test results, this file contains details about source code coverage by tests that are calculated in three levels.

> *   Project coverage

Reviewer 2: ```suggestion
*   Package coverage
```

> *   Module coverage

Reviewer 2: ```suggestion
*   Individual module coverage
```
